### ISSUE TITLE
Local version causes mismatch between candidate and installed package

### ISSUE THREAD

**Issue by tgolsson** [pdm-project/pdm#1497]
- [x] I have searched the issue tracker and believe that this is not a duplicate.

**Make sure you run commands with `-v` flag before pasting the output.**

## Context

This is the reduced case of #1496, but which is actually two different issues masquerading to look like a cache issue. 

## Steps to reproduce

Repeatedly `pdm sync -v` here and note that each time the torch package is deleted and reinstalled:

https://github.com/tgolsson/pdm-repros/tree/main/constant-redownload

## Actual behavior

The same package is resolved (and reinstalled) each time when it's bitwise the same.  This also leads to `unearth` claiming to download the package, which is what sent me down the cache path.

Note that `pdm list` recognizes the actual version installed:
```
╭─────────────────────┬────────────┬─────────────────────────────────────────────────────────╮
│ name                │ version    │ location                                                │
├─────────────────────┼────────────┼─────────────────────────────────────────────────────────┤
│ setuptools          │ 59.5.0     │                                                         │
│ torch               │ 1.10.2+cpu │                                                         │
│ wasteful-redownload │ 0.0.0      │ -e /home/ts/Repositories/pdm-repros/constant-redownload │
│ typing_extensions   │ 4.4.0      │                                                         │
╰─────────────────────┴────────────┴─────────────────────────────────────────────────────────╯
```

## Expected behavior

Something like this is my expected behavior:

| Requirement         | Installed           | Expected outcome |
| ------------------- | ------------------- | ---------------- |
| `torch==1.10.2`     | `torch==1.10.2`     | Keep             |
| `torch==1.10.2`     | `torch==1.10.2+cpu` | Keep             |
| `torch==1.10.2+cpu` | `torch==1.10.2`     | Update           |
| `torch==1.10.2+cpu` | `torch==1.10.2+cpu` | Keep             |
| `torch==1.10.2+cpu` | `torch==1.10.2+gpu` | Update           |

## Environment Information

```bash
# Paste the output of `pdm info && pdm info --env` below:
# Paste the output of `pdm info && pdm info --env` below:
PDM version:
  2.2.1
Python Interpreter:
  /home/ts/Repositories/pdm-repros/constant-redownload/.venv/bin/python (3.8)
Project Root:
  /home/ts/Repositories/pdm-repros/constant-redownload
Project Packages:
  None
{
  "implementation_name": "cpython",
  "implementation_version": "3.8.10",
  "os_name": "posix",
  "platform_machine": "x86_64",
  "platform_release": "5.15.68.1-microsoft-standard-WSL2",
  "platform_system": "Linux",
  "platform_version": "#1 SMP Mon Sep 19 19:14:52 UTC 2022",
  "python_full_version": "3.8.10",
  "platform_python_implementation": "CPython",
  "python_version": "3.8",
  "sys_platform": "linux"
}
```

**Comment by frostming** [comment]
After a little research I found that

- The reinstallation isn't expected and should be fixed
- The cache mismatch is due to lacking of `expires` or `max-age` header.

**Comment by tgolsson** [comment]
@frostming Yepp! I put up a PR to avoid the cache thrashing caused by the headers.

**Comment by frostming** [comment]
This was fixed in e480d2c, the cache issue is tracked by #1498
